File: service/src/main/java/com/github/goldy1992/mp3player/service/MediaSessionConnectorCreator.java
package com.github.goldy1992.mp3player.service;

import android.support.v4.media.session.MediaSessionCompat;

import com.github.goldy1992.mp3player.service.player.DecreaseSpeedProvider;
import com.github.goldy1992.mp3player.service.player.IncreaseSpeedProvider;
import com.github.goldy1992.mp3player.service.player.MyMediaButtonEventHandler;
import com.github.goldy1992.mp3player.service.player.MyMetadataProvider;
import com.github.goldy1992.mp3player.service.player.MyPlaybackPreparer;
import com.github.goldy1992.mp3player.service.player.MyTimelineQueueNavigator;
import com.google.android.exoplayer2.ExoPlayer;
import com.google.android.exoplayer2.ext.mediasession.MediaSessionConnector;

import javax.inject.Inject;
import javax.inject.Singleton;

import static android.support.v4.media.session.PlaybackStateCompat.ACTION_PAUSE;
import static android.support.v4.media.session.PlaybackStateCompat.ACTION_PLAY;
import static android.support.v4.media.session.PlaybackStateCompat.ACTION_SEEK_TO;
import static android.support.v4.media.session.PlaybackStateCompat.ACTION_SET_REPEAT_MODE;
import static android.support.v4.media.session.PlaybackStateCompat.ACTION_SET_SHUFFLE_MODE;
import static android.support.v4.media.session.PlaybackStateCompat.ACTION_STOP;


@Singleton
public final class MediaSessionConnectorCreator {

    @MediaSessionConnector.PlaybackActions
    public static final long SUPPORTED_PLAYBACK_ACTIONS = ACTION_STOP | ACTION_PAUSE | ACTION_PLAY |
            ACTION_SET_REPEAT_MODE | ACTION_SET_SHUFFLE_MODE | ACTION_SEEK_TO;



    private MediaSessionConnector mediaSessionConnector = null;
    private final MediaSessionCompat mediaSessionCompat;
    private final ExoPlayer exoPlayer;
    private final MyPlaybackPreparer myPlaybackPreparer;
    private final MyControlDispatcher myControlDispatcher;
    private final MyMetadataProvider myMetadataProvider;
    private final MyTimelineQueueNavigator myTimelineQueueNavigator;
    private final IncreaseSpeedProvider increaseSpeedProvider;
    private final DecreaseSpeedProvider decreaseSpeedProvider;
    private final MyMediaButtonEventHandler myMediaButtonEventHandler;

    @Inject
    public MediaSessionConnectorCreator( MediaSessionCompat mediaSessionCompat,
                                         ExoPlayer exoPlayer,
                                         MyPlaybackPreparer myPlaybackPreparer,
                                         MyControlDispatcher myControlDispatcher,
                                         MyMetadataProvider myMetadataProvider,
                                         MyTimelineQueueNavigator myTimelineQueueNavigator,
                                         IncreaseSpeedProvider increaseSpeedProvider,
                                         DecreaseSpeedProvider decreaseSpeedProvider,
                                         MyMediaButtonEventHandler myMediaButtonEventHandler) {
        this.mediaSessionCompat = mediaSessionCompat;
        this. exoPlayer = exoPlayer;
        this.myPlaybackPreparer = myPlaybackPreparer;
        this.myControlDispatcher = myControlDispatcher;
        this.myMetadataProvider = myMetadataProvider;
        this.myTimelineQueueNavigator = myTimelineQueueNavigator;
        this.increaseSpeedProvider = increaseSpeedProvider;
        this.decreaseSpeedProvider = decreaseSpeedProvider;
        this.myMediaButtonEventHandler = myMediaButtonEventHandler;
    }

    public MediaSessionConnector create() {
        if (null == this.mediaSessionConnector) {
            MediaSessionConnector newMediaSessionConnector = new MediaSessionConnector(mediaSessionCompat);
            newMediaSessionConnector.setPlayer(exoPlayer);
            newMediaSessionConnector.setPlaybackPreparer(myPlaybackPreparer);
            newMediaSessionConnector.setControlDispatcher(myControlDispatcher);
            newMediaSessionConnector.setMediaMetadataProvider(myMetadataProvider);
            newMediaSessionConnector.setQueueNavigator(myTimelineQueueNavigator);
            newMediaSessionConnector.setCustomActionProviders(increaseSpeedProvider, decreaseSpeedProvider);
            newMediaSessionConnector.setMediaButtonEventHandler(myMediaButtonEventHandler);
            newMediaSessionConnector.setEnabledPlaybackActions(SUPPORTED_PLAYBACK_ACTIONS);
            this.mediaSessionConnector = newMediaSessionConnector;
        }
        return this.mediaSessionConnector;
    }
}

File: service/src/main/java/com/github/goldy1992/mp3player/service/MediaSessionConnectorCreator.kt
package com.github.goldy1992.mp3player.service

import android.os.Bundle
import android.support.v4.media.session.MediaSessionCompat
import android.support.v4.media.session.PlaybackStateCompat
import com.github.goldy1992.mp3player.commons.Constants.AUDIO_SESSION_ID
import com.github.goldy1992.mp3player.service.player.*
import com.google.android.exoplayer2.ExoPlayer
import com.google.android.exoplayer2.Player
import com.google.android.exoplayer2.ext.mediasession.MediaSessionConnector
import com.google.android.exoplayer2.ext.mediasession.MediaSessionConnector.PlaybackActions
import dagger.hilt.android.scopes.ServiceScoped
import javax.inject.Inject

@ServiceScoped
class MediaSessionConnectorCreator
    @Inject
    constructor(
        private val mediaSessionCompat: MediaSessionCompat,
        private val exoPlayer: ExoPlayer,
        private val myPlaybackPreparer: MyPlaybackPreparer,
     //   private val myMetadataProvider: MyMetadataProvider,
        private val myTimelineQueueNavigator: MyTimelineQueueNavigator,
        private val changeSpeedProvider: ChangeSpeedProvider,
        private val myMediaButtonEventHandler: MyMediaButtonEventHandler,
        private val playlistManager: PlaylistManager) {
    private var mediaSessionConnector: MediaSessionConnector? = null
    fun create(): MediaSessionConnector? {
        if (null == mediaSessionConnector) {

            val newMediaSessionConnector = MediaSessionConnector(mediaSessionCompat)
            mediaSessionConnector = newMediaSessionConnector
            val toSend = Bundle()
            toSend.putInt(AUDIO_SESSION_ID, exoPlayer.audioSessionId)
            mediaSessionCompat.setExtras(toSend)

            if (!playlistManager.isEmpty()) {
                newMediaSessionConnector.setPlaybackPreparer(myPlaybackPreparer)
                newMediaSessionConnector.setMediaMetadataProvider(MediaSessionConnector.DefaultMediaMetadataProvider(mediaSessionCompat.controller, null))
                newMediaSessionConnector.setQueueNavigator(myTimelineQueueNavigator)
                newMediaSessionConnector.setCustomActionProviders(changeSpeedProvider)
                newMediaSessionConnector.setMediaButtonEventHandler(myMediaButtonEventHandler)
                newMediaSessionConnector.setEnabledPlaybackActions(SUPPORTED_PLAYBACK_ACTIONS)
                newMediaSessionConnector.setPlayer(exoPlayer)
            }
        }
        return mediaSessionConnector
    }

    companion object {
        @PlaybackActions
        val SUPPORTED_PLAYBACK_ACTIONS =
            PlaybackStateCompat.ACTION_STOP or
            PlaybackStateCompat.ACTION_PAUSE or
            PlaybackStateCompat.ACTION_PLAY or
            PlaybackStateCompat.ACTION_SET_REPEAT_MODE or
            PlaybackStateCompat.ACTION_SET_SHUFFLE_MODE or
            PlaybackStateCompat.ACTION_SEEK_TO
    }

}